operations/helm/charts/mimir-distributed/templates/kafka/kafka-statefulset.yaml

```diff
@@ -0,0 +1,175 @@
+  labels:
+    {{- include "mimir.labels" (dict "ctx" $ "component" "kafka") | nindent 4 }}
+  annotations:
+    {{- toYaml .podAnnotations | nindent 4 }}
+  namespace: {{ $.Release.Namespace | quote }}
+spec:
+  podManagementPolicy: Parallel
+  replicas: {{ $replicas }}
+  selector:
+    matchLabels:
+      {{- include "mimir.selectorLabels" (dict "ctx" $ "component" "kafka") | nindent 6 }}
+  serviceName: {{ include "mimir.resourceName" (dict "ctx" $ "component" "kafka") }}-headless
+  template:
+    metadata:
+      labels:
+        {{- include "mimir.podLabels" (dict "ctx" $ "component" "kafka") | nindent 8 }}
+      annotations:
+        {{- include "mimir.podAnnotations" (dict "ctx" $ "component" "kafka") | nindent 8 }}
+      namespace: {{ $.Release.Namespace | quote }}
+    spec:
+      serviceAccountName: {{ template "mimir.serviceAccountName" $ }}
+      {{- if .priorityClassName }}
+      priorityClassName: {{ .priorityClassName }}
+      {{- end }}
+      securityContext:
+        {{- toYaml .securityContext | nindent 8 }}
+      {{- if .image.pullSecrets }}
+      imagePullSecrets:
+      {{- range .image.pullSecrets }}
+        - name: {{ . }}
+      {{- end }}
+      {{- end }}
+      {{- with .nodeSelector }}
+      nodeSelector:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      {{- with .affinity }}
+      affinity:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      {{- with .tolerations }}
+      tolerations:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      terminationGracePeriodSeconds: {{ .terminationGracePeriodSeconds }}
+      containers:
+        - name: kafka
+          image: "{{ .image.repository }}:{{ .image.tag }}"
+          imagePullPolicy: {{ .image.pullPolicy }}
+          ports:
+            - containerPort: 9092
+              name: kafka
+              protocol: TCP
+            - containerPort: 9093
+              name: controller
+              protocol: TCP
+          env:
+            # Kubernetes expands env variables in lexical order. So "_POD_NAME" (sic) here is so the name was expanded before KAFKA_ADVERTISED_LISTENERS is.
+            - name: _POD_NAME
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.name
+            - name: KAFKA_CLUSTER_ID
+              value: "{{ .clusterId | default "" }}"
+            - name: KAFKA_NODE_ID
+              valueFrom:
+                fieldRef:
+                  fieldPath: metadata.labels['apps.kubernetes.io/pod-index']
+            - name: KAFKA_PROCESS_ROLES
+              value: "broker,controller"
+            - name: KAFKA_LISTENERS
+              value: "PLAINTEXT://0.0.0.0:9092,CONTROLLER://0.0.0.0:9093"
+            - name: KAFKA_ADVERTISED_LISTENERS
+              value: PLAINTEXT://$(_POD_NAME).{{ include "mimir.resourceName" (dict "ctx" $ "component" "kafka") }}-headless.{{ $.Release.Namespace }}.svc:9092
```

Comment:
you can use `.values.global.clusterDomain` here for an fqdn

Reply:
and if you do that can we drop the `KAFKA_CONTROLLER_QUORUM_VOTERS`

Reply:
> you can use .values.global.clusterDomain here for an fqdn

It works now. I don't understand why it didn't yesterday. Done in https://github.com/grafana/mimir/pull/12459/commits/fb2b6ee0532ace62637ac20304b5534373fd3949

Reply:
> do we still need KAFKA_CONTROLLER_QUORUM_VOTERS then?

Yes. Otherwise it dies on startup,

```
===> Configuring ...
Running in KRaft mode...
===> Launching ...
Exception in thread "main" org.apache.kafka.common.config.ConfigException: If using process.roles, either controller.quorum.bootstrap.servers must contain the set of bootstrap controllers or controller.quorum.voters must contain a parseable set of controllers.
```